File: RPC/2020:06:13 Junio/editorial/errorcorrection/submissions/accepted/errorcorrection_hamerly_hk.py
#!/usr/bin/env python3

import collections
import sys

# solution to errorcorrection, using bipartite matching
#
# - Create a graph like the following, with edges between words that are
#   swappable. These edges represent pairs of words that cannot be in a
#   swap-free set together.
#
# - Determine which nodes have even and odd permutation parity
#   - With this parity concept, the graph is bipartite.
#
# - Find a maximum bipartite matching using the Hopcroft-Karp algorithm
#   - This matching represents the largest group of word pairs that cannot
#     appear together. If there were more word pairs that could be matched, the
#     matching would be larger.
#
# - The largest swap-free set is the number of original words minus the flow in
#   the network (which is the number of matched word pairs).
#
# This version uses Hopcroft-Karp, which should be permissible.

def parity(s):
    '''Define the parity of a sequence as the number of inversions in the
    sequence, which we calculate brute-force. Two words with equal parity
    cannot be swappable; two words with opposing parity may be swappable. Thus
    the two parities form the two sides of our bipartite graph.'''
    p = 0
    for i, c in enumerate(s):
        for j, c2 in enumerate(s[i+1:], i+1):
            if c2 < c:
                p += 1
    return p % 2

oo = 1e100

SINK = None

def bfs(edges, fwd, back, dist):
    frontier = collections.deque()
    for u in fwd:
        if fwd[u] is None:
            dist[u] = 0
            frontier.append(u)
        else:
            dist[u] = oo

    # None here indicates the sink we're looking for
    dist[SINK] = oo

    while frontier:
        u = frontier.popleft()
        if dist[u] < dist[SINK]:
            for v in edges[u]:
                # NB: back[v] could be None, indicating it is an endpoint; or
                # it could be another vertex among those in the starting
                # portion of the graph
                if dist[back[v]] == oo:
                    dist[back[v]] = dist[u] + 1
                    frontier.append(back[v])

    return dist[SINK] != oo

def dfs(edges, u, fwd, back, dist):
    if u == SINK:
        return True

    for v in edges[u]:
        if dist[back[v]] == dist[u] + 1:
            if dfs(edges, back[v], fwd, back, dist):
                fwd[u] = v
                back[v] = u
                return True

    return False

def swap(s, i, j):
    l = list(s)
    l[i], l[j] = l[j], l[i]
    return ''.join(l)

def main():
    n = int(sys.stdin.readline())
    words = [sys.stdin.readline().strip() for _ in range(n)]
    word_index = {w: i for i, w in enumerate(words)}
    edges = {i: [] for i in range(n)}

    # compute the parity of the permutation of each word -- e.g. the parity of
    # the number of inversions are in the string. Connect words that have
    # opposing parity and are one-swap distant.
    parity_words = {0: [], 1: []}
    k = len(words[0])
    for i, w in enumerate(words):
        p = parity(w)
        parity_words[p].append(i)

        # if the parity is 0, then find out if there are any one-swap neighbors
        # that are also in the word list
        if p == 0:
            for a in range(k):
                for b in range(a+1,k):
                    s = swap(w, a, b)
                    if s in word_index:
                        edges[i].append(word_index[s])

    # parings
    fwd = {i: None for i in parity_words[0]}
    back = {i: None for i in parity_words[1]}
    dist = {i: None for i in range(n)}

    # Hopcroft-Karp
    while bfs(edges, fwd, back, dist):
        for u in fwd:
            if fwd[u] is None:
                dfs(edges, u, fwd, back, dist)

    total_flow = sum(v is not None for u, v in fwd.items())
    print(n - total_flow)

if __name__ == '__main__':
    main()

File: RPC/2020:06:13 Junio/editorial/errorcorrection/submissions/accepted/errorcorrection_hamerly_hk.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>
#include <unordered_map>
#include <cassert>

int parity(std::string const &s) {
    bool p = false;
    for (auto i = s.begin(); i != s.end(); ++i)
        for (auto j = i + 1; j != s.end(); ++j)
            if (*j < *i)
                p = !p;
    return (int)p;
}

std::string swap(std::string const &s, size_t i, size_t j) {
    std::string w(s);
    w[i] = s[j];
    w[j] = s[i];
    return w;
}

int const SINK = -1;
int const oo = 10000000;

typedef std::vector<std::vector<int>> Graph;

bool bfs(Graph const &edges, std::map<int, int> const &fwd, std::map<int, int> const &back, std::map<int, int> &dist) {
    std::queue<int> frontier;

    for (auto [u, v] : fwd) {
        if (v == -1) {
            dist[u] = 0;
            frontier.push(u);
        } else {
            dist[u] = oo;
        }
    }

    dist[SINK] = oo;

    while (frontier.size()) {
        int u = frontier.front();
        frontier.pop();

        if (dist[u] < dist[SINK]) {
            for (auto v : edges[u]) {
                auto bv = back.find(v);
                assert(bv != back.end());
                if (dist[bv->second] == oo) {
                    dist[bv->second] = dist[u] + 1;
                    frontier.push(bv->second);
                }
            }
        }
    }

    return dist[SINK] != oo;
}

bool dfs(Graph const &edges, int u, std::map<int, int> *fwd, std::map<int, int> *back, std::map<int, int> const &dist) {
    if (u == SINK)
        return true;

    for (auto v : edges[u]) {
        if (dist.find((*back)[v])->second == dist.find(u)->second + 1) {
            if (dfs(edges, (*back)[v], fwd, back, dist)) {
                (*fwd)[u] = v;
                (*back)[v] = u;
                return true;
            }
        }
    }

    return false;
}

int main() {
    int n;
    std::cin >> n;

    std::vector<std::string> words(n);
    std::unordered_map<std::string, size_t> word_index;
    for (int i = 0; i < n; ++i) {
        std::cin >> words[i];
        word_index[words[i]] = i;
    }

    Graph edges(n);

    std::vector<std::vector<int>> parity_words(2);

    size_t k = words[0].size();
    for (int i = 0; i < n; ++i) {
        int p = parity(words[i]);
        parity_words[p].push_back(i);

        if (p == 0) {
            for (size_t a = 0; a < k; ++a) {
                for (size_t b = a + 1; b < k; ++b) {
                    auto s = swap(words[i], a, b);
                    auto si = word_index.find(s);
                    if (si != word_index.end())
                        edges[i].push_back(si->second);
                }
            }
        }
    }

    std::map<int, int> fwd, back, dist;

    for (auto u : parity_words[0]) fwd[u] = -1;
    for (auto v : parity_words[1]) back[v] = -1;
    for (int u = 0; u < n; ++u) dist[u] = oo;

    while (bfs(edges, fwd, back, dist)) {
        for (auto [u, v] : fwd) {
            if (v == -1)
                dfs(edges, u, &fwd, &back, dist);
        }
    }

    int total_flow = 0;
    for (auto [u, v] : fwd)
        total_flow += (v != -1);
    std::cout << (n - total_flow) << std::endl;

    return 0;
}

